crypto/bio/bss_mem.c

```diff
@@ -193,17 +253,43 @@ static int mem_read(BIO *b, char *out, int outl)
     int ret = -1;
     BIO_BUF_MEM *bbm = (BIO_BUF_MEM *)b->ptr;
     BUF_MEM *bm = bbm->readp;
+    size_t maxreadlen = 0;
+    int eof = 0;
 
     if (b->flags & BIO_FLAGS_MEM_RDONLY)
         bm = bbm->buf;
     BIO_clear_retry_flags(b);
-    ret = (outl >= 0 && (size_t)outl > bm->length) ? (int)bm->length : outl;
+    if (bbm->use_dgrams) {
+        if (bbm->dgrams != NULL) {
+            maxreadlen = bbm->dgrams->dgramlen;
+            if (!ossl_assert(maxreadlen <= bm->length))
+                return 0;
+        } else {
+            eof = 1;
+        }
+    } else {
+        maxreadlen = bm->length;
+        eof = (maxreadlen == 0);
+    }
+    ret = (outl >= 0 && (size_t)outl > maxreadlen) ? (int)maxreadlen : outl;
     if ((out != NULL) && (ret > 0)) {
+        size_t flushlen;
+
         memcpy(out, bm->data, ret);
-        bm->length -= ret;
-        bm->max -= ret;
-        bm->data += ret;
-    } else if (bm->length == 0) {
+        flushlen = bbm->use_dgrams ? maxreadlen : (size_t)ret;
+            
+        bm->length -= flushlen;
+        bm->max -= flushlen;
+        bm->data += flushlen;
+        if (bbm->use_dgrams) {
+            struct buf_mem_dgram_st *tmp = bbm->dgrams;
+
+            bbm->dgrams = tmp->next;
```

Comment:
I don't like the read what fits and discard the rest semantics here.  It just seems wrong.

I'd suggest either:
* return all the data that fits and the next read continues with the **same** datagram; or
* return an error, reading nothing.

Reply:
We have to have this behaviour to match the semantics of the Berkeley sockets API.

I don't like this behaviour either, which is why my dgram_pair supports a new BIO_dgram_set_no_trunc() call to cause reads with an inadequate buffer to fail. But I don't think we can make this the default.

Reply:
Do you think the `return an error, reading nothing` option would be incompatible with the sockets API? IMO this is within its spec.

Reply:
Return an error, reading nothing is the behaviour enabled by `BIO_dgram_set_no_trunc`, so it's an option. Since this BIO is probably going to be primarily used internally I don't see a need to implement it unless @mattcaswell wants the functionality for libssl internals.

Reply:
I don't need it for libssl internals.